src/libraries/System.Net.Http/src/System/Threading/AsyncMutex.cs

```diff
@@ -0,0 +1,313 @@
+    /// where the mutex is held for a very short period of time, when it is held any other
+    /// attempts to access it must wait asynchronously, where it's only binary rather than counting, and where
+    /// we want to minimize contention that a releaser incurs while trying to unblock a waiter.  The primary
+    /// value-add is the fast-path interlocked checks that minimize contention for these use cases (essentially
+    /// making it an async futex), and then as long as we're wrapping something and we know exactly how all
+    /// consumers use the type, we can offer a ValueTask-based implementation that reuses waiter nodes.
+    /// </remarks>
+    internal sealed class AsyncMutex
+    {
+        /// <summary>Fast-path gate count tracking access to the mutex.</summary>
+        /// <remarks>
+        /// If the value is 1, the mutex can be entered atomically with an interlocked operation.
+        /// If the value is less than or equal to 0, the mutex is held and requires fallback to enter it.
+        /// </remarks>
+        private int _gate = 1;
+        /// <summary>Secondary check guarded by the lock to indicate whether the mutex is acquired.</summary>
+        /// <remarks>
+        /// This is only meaningful after having updated <see cref="_gate"/> via interlockeds and taken the appropriate path.
+        /// If after decrementing <see cref="_gate"/> we end up with a negative count, the mutex is contended, hence
+        /// <see cref="_lockedSemaphoreFull"/> starting as <c>true</c>.  The primary purpose of this field
+        /// is to handle the race condition between one thread acquiring the mutex, then another thread trying to acquire
+        /// and getting as far as completing the interlocked operation, and then the original thread releasing; at that point
+        /// it'll hit the lock and we need to store that the mutex is available to enter.  If we instead used a
+        /// SemaphoreSlim as the fallback from the interlockeds, this would have been its count, and it would have started
+        /// with an initial count of 0.
+        /// </remarks>
+        private bool _lockedSemaphoreFull = true;
+        /// <summary>The head of the double-linked waiting queue.  Waiters are dequeued from the head.</summary>
+        private Waiter? _waitersHead;
+        /// <summary>The tail of the double-linked waiting queue.  Waiters are added at the tail.</summary>
+        private Waiter? _waitersTail;
+        /// <summary>A pool of waiter objects that are ready to be reused.</summary>
+        /// <remarks>
+        /// There is no bound on this pool, but it ends up being implicitly bounded by the maximum number of concurrent
+        /// waiters there ever were, which for our uses in HTTP/2 will end up being the high-water mark of concurrent streams
+        /// on a single connection.
+        /// </remarks>
+        private readonly ConcurrentQueue<Waiter> _unusedWaiters = new ConcurrentQueue<Waiter>();
```

Comment:
Instead of using a separate queue here, you could make a simple linked list using `Waiter.Next`

Reply:
What did you have in mind?  This is accessed concurrently.  Part of my goal was to remove as much from the lock as possible.

Reply:
Alas without https://github.com/dotnet/runtime/issues/31911 the enqueue would not be lock-free.